Added parallel mode to utilise power of all CPUs

doterax opened the pull request:
Taskflow added from sources to make AppVeyor happy
Changed c++14 to c++17 to support taskflow
Added option -p --parallel to enable parallel mode
Logs changed to prevent mess in it

JayXon commented on main.cpp:
```diff
@@ -43,28 +55,39 @@ int ProcessFile(const char* file_path, const struct stat* sb = nullptr, int type
-    PrintSize(original_size - new_size);
+    std::string log;
+    if (concurrent_processing)
+      log = "Processing: " + filename + "\n";
```
nit: since with concurrent this will be printed afterwards, maybe we should print something different, Processed? Optimized? Leanified?

doterax replied:
Word "Processed" used in parallel mode.

JayXon commented on main.cpp:
```diff
@@ -221,9 +274,17 @@ int main(int argc, char** argv) {
+  size_t concurrent_tasks = std::thread::hardware_concurrency();
+
+  if (!concurrent_processing)
+    concurrent_tasks = 1;
```
I'd reverse this, only call `hardware_concurrency()` if `concurrent_processing`

doterax replied:
done

JayXon commented on main.cpp:
```diff
@@ -16,20 +16,32 @@
 
+
 #ifdef _WIN32
 int ProcessFile(const wchar_t* file_path) {
```
now that this is wrapped by `EnqueueProcessFileTask`, we can change it to take `const string&` and move all the ifdef platform specific stuff to `EnqueueProcessFileTask`

doterax replied:
We can't get rid of ifdefs because for windows we need wchar_t* for filenames and for the rest char* used. But I removed ftw() stuff from ProcessFile.

JayXon commented on main.cpp:
```diff
@@ -173,6 +223,9 @@ int main(int argc, char** argv) {
           } else if (STRCMP(argv[i] + j + 1, "verbose") == 0) {
             j += 6;
             argv[i][j + 1] = 'v';
+          } else if (STRCMP(argv[i] + j + 1, "concurrent") == 0) {
```
I think parallel is a better name for this
https://softwareengineering.stackexchange.com/questions/190719/the-difference-between-concurrent-and-parallel-execution

doterax replied:
Reworded for 'parallel' :)

JayXon reviewed (changes requested):
Thanks for the pull request! Overall looks pretty good, a couple things:

### taskflow 
Could you move all the taskflow files into the `lib` directory? That's where all the third party libraries live.
Also since we are checking in code into this repo, we should remove files not needed, for example we probably don't need any of those cuda or tensor files...

### AppVeyor failures
Clang-Release; Platform: x86 you missed one of the `<ExceptionHandling>false</ExceptionHandling>` lines
Clang-Release; Platform: x64 this seems to be something else, ~~I'm not sure why it's complaining, but doesn't look like your fault, I'll see if I can fix this.~~ fixed in 7075d27

### Travis CI failures
I'm not sure why it doesn't show up in this PR, but it's building this and all failed: https://travis-ci.org/github/JayXon/Leanify/builds/762323769

### Logs
I see that you've reordered one of the logs, but that doesn't address any verbose logs, (try `-c -v`), do you have any plans for those?

doterax commented:
Thank you for review!

I processed all your requests. Except removing not needed files from taskflow. That because I have some plans for them.

Looks like problems with linking in Clang configurations. I will check them tomorrow on a different machine. Or if you have suggestions don't hesitate to contact me. 

Cheers :)

JayXon reviewed (commented):
> I processed all your requests. Except removing not needed files from taskflow. That because I have some plans for them.

Interesting, do you mind sharing the plans?

> Looks like problems with linking in Clang configurations. I will check them tomorrow on a different machine. Or if you have suggestions don't hesitate to contact me.

Probably missing pthread?

doterax commented on main.cpp:
```diff
@@ -48,14 +48,11 @@ int ProcessFile(const wchar_t* file_path) {
   WideCharToMultiByte(CP_ACP, 0, file_path, -1, mbs, sizeof(mbs) - 1, nullptr, nullptr);
```
`WideCharToMultiByte` is needed only for non unicode console output in `ProcessFile()` when platform is WIN32. 

So as result we should have both string types for windows: `const wchar_t*` for File constructor `const char*` for console out. And I think `ProcessFile()` is the right place for `WideCharToMultiByte`

I made small changes and prevent string creation and copy for non WIN32 code.

doterax commented:
> Interesting, do you mind sharing the plans?

Yes, sorry.

My future plans is:
	- change the way how comandline processed
	- add more commandline options for zopflipng (for example diffirentiate number of iterations for small and big files for zopfli)
	- try playing with guetzli-cuda-opencl for JPEG (https://github.com/ianhuang-777/guetzli-cuda-opencl/tree/opencl)

doterax commented:
Travis CI build failed due to not full support of C++17.

https://stackoverflow.com/questions/40177006/c-variant-no-such-file-or-directory

JayXon commented on main.cpp:
```diff
@@ -16,7 +16,7 @@
 #include "version.h"
 
-#include "taskflow/taskflow.hpp"
+#include "lib/taskflow/taskflow.hpp"
```
Could you address this? I don't see any new commits

doterax replied:
fixed

JayXon commented:
> - try playing with guetzli-cuda-opencl for JPEG (https://github.com/ianhuang-777/guetzli-cuda-opencl/tree/opencl)

I don't think we can use guetzli in leanify, because leanify is a lossless tool but guetzli is lossy.

> Travis CI build failed due to not full support of C++17.

Sorry about this, I've already switched from travis ci to github actions, it should show up if you add a new commit.

doterax commented:
> Could you address this? I don't see any new commits

Yes, sorry I set up system to reproduce Clang linking issues and didn't commit before fix. And now AppVeyor works.

I had to disable optimisation for Clang, and this helped. 
Someone also faced same issue https://github.com/taskflow/taskflow/issues/300

> I don't think we can use guetzli in leanify, because leanify is a lossless tool but guetzli is lossy.

Ok, I understand. I removed all unused taksflow files.

> I've already switched from travis ci to github actions, it should show up if you add a new commit.

`gcc build on ubuntu-16.04` still not fully support C++17

Rest configurations on ubuntu are failed, and for now I have no idea how to fix them. Because problem looks same as in Clang, I tried to remove optimisation from makefile, but this is not worked.

JayXon commented on main.cpp:
```diff
@@ -9,62 +9,84 @@
 #include <ftw.h>
 #endif
 
+#include <taskflow/taskflow.hpp>
```
one last nit: add an empty line between library header and local header

doterax replied:
done

JayXon reviewed (commented):
> I had to disable optimisation for Clang, and this helped.
> Someone also faced same issue [taskflow/taskflow#300](https://github.com/taskflow/taskflow/issues/300)

Hmm I think we should figure out why it doesn't work with optimizations, right now this is a regression for using leanify on single file, at least I want to know exactly which optimization is causing this issue, do we really need to disable all the optimizations?

> `gcc build on ubuntu-16.04` still not fully support C++17

Excluded that from the matrix

> Rest configurations on ubuntu are failed, and for now I have no idea how to fix them. Because problem looks same as in Clang, I tried to remove optimisation from makefile, but this is not worked.

Try adding `-pthread` to the makefile?

doterax commented:
> Try adding -pthread to the makefile?

It works. Thank you.

> Do we really need to disable all the optimizations?

I'm trying to figure out why it helps and how to preserve all optimizations.

JayXon commented on Makefile:
```diff
@@ -5,7 +5,7 @@ PUGIXML_OBJ     := lib/pugixml/pugixml.o
 ZOPFLIPNG_OBJ   := lib/zopflipng/lodepng/lodepng.o lib/zopflipng/lodepng/lodepng_util.o lib/zopflipng/zopflipng_lib.o
 
-CFLAGS      += -Wall -Wextra -Wno-unused-parameter -Werror -O3 -msse2 -mfpmath=sse -flto
+CFLAGS      += -Wall -Wextra -Wno-unused-parameter -Werror -O3 -msse2 -mfpmath=sse -flto -pthread
```
This should be added to LDFLAGS instead because it's only used during linking.

doterax replied:
Fixed.

doterax commented:
> I'm trying to figure out why it helps and how to preserve all optimizations.

I'm working on it. Looks like rare LLVM bug. Have plans to rewrite TLS part of taskflow an enable optimisation for Clang. Also somehow SEH on Windows also involved.

doterax commented:
Looks like everything is green now :)

JayXon commented:
Thanks!
